### PR TITLE
bin, etc: support config file.

### PR DESCRIPTION
**Pull request by overvenus**
see #248

### CODE REVIEW

File: src/bin/tikv-server.rs [hunk 1]
```diff
@@ -86,8 +89,22 @@ fn build_raftkv(matches: &Matches,
...
     // Set advertise address for outer node and client use.
     // If no advertise listening address set, use the associated listening address.
-    cfg.advertise_addr = matches.opt_str("advertise-addr")
-                                .unwrap_or_else(|| addr);
+    cfg.advertise_addr = matches.opt_str("advertise-addr").unwrap_or_else(|| {
+        match config.lookup("server.advertise-addr") {
+            Some(v) => {
+                v.as_str()
+                 .unwrap_or_else(|| {
+                     info!("mlfromed advertise-addr, use addr: {}", addr);
```
Comment (siddontang) [inline src/bin/tikv-server.rs]: malformed

File: src/bin/tikv-server.rs [hunk 2]
```diff
@@ -96,14 +113,23 @@ fn build_raftkv(matches: &Matches,
     (create_raft_storage(node, engine).unwrap(), raft_router)
 }
 
-fn get_store_path(matches: &Matches) -> String {
-    let path = matches.opt_str("s");
-    if path.is_none() {
-        return TEMP_DIR.to_owned();
-    }
+fn get_store_path(matches: &Matches, config: &toml::Value) -> String {
+    let path = match matches.opt_str("s") {
```
Comment (siddontang) [inline src/bin/tikv-server.rs]: why not using unwrap_or_else here?

Comment (overvenus) [inline src/bin/tikv-server.rs]: `unwrap_or_else<F: FnOnce() -> T>(self, f: F) -> T` takes a function

e.g.:

```rust
foo.unwrap_or_else(|| {
    return TEMP_DIR.to_owned(),
});
```

but here wants return `TEMP_DIR` immediately if store was unspecified.

**Comment by ngaut** [comment]
PTAL @siddontang @BusyJay

File: src/bin/tikv-server.rs [hunk 3]
```diff
@@ -20,14 +20,15 @@ extern crate getopts;
 extern crate log;
 extern crate rocksdb;
 extern crate mio;
+extern crate toml;
 
 use std::env;
 use std::fs;
 use std::path::Path;
 use std::sync::{Arc, RwLock};
+use std::io::prelude::Read;
```
Comment (BusyJay) [inline src/bin/tikv-server.rs]: import `std::io::Read` instead.

File: src/bin/tikv-server.rs [hunk 4]
```diff
@@ -47,24 +48,26 @@ fn print_usage(program: &str, opts: Options) {
...
 }
 
-fn initial_log(matches: &Matches) {
-    let log_filter = match matches.opt_str("L") {
-        Some(level) => logger::get_level_by_string(&level),
-        None => LogLevelFilter::Info,
-    };
-    util::init_log(log_filter).unwrap();
+fn initial_log(matches: &Matches, config: &toml::Value) {
+    let level = matches.opt_str("L").unwrap_or_else(|| {
```
Comment (BusyJay) [inline src/bin/tikv-server.rs]: Use `or_else` chain seems more neat.

File: src/bin/tikv-server.rs [hunk 5]
```diff
@@ -181,24 +217,66 @@ fn main() {
...
+            config_file.read_to_string(&mut s).expect("config read filed");
+            toml::Value::Table(toml::Parser::new(&s).parse().expect("malformed config file"))
+        }
+        // Empty value, lookup() always return `None`.
+        None => toml::Value::Integer(0),
+    };
+
+    initial_log(&matches, &config);
+
+    let addr = matches.opt_str("A").unwrap_or_else(|| {
```
Comment (BusyJay) [inline src/bin/tikv-server.rs]: I think these similar code snippets can be extracted to a function.

Comment (overvenus) [inline src/bin/tikv-server.rs]: I had tried, but found it is difficult to implement.

1. Different values have different types,  `toml::Value::{String, Integer, Bool ...}`, which lead to `as_str(), as_integer(), as_bool() ...`

2. Different values have different behaves, eg: get `addr` from: command line > config file > default; get `pd` from: command line > config file > **panic!()** (no default value for pd)

Comment (BusyJay) [inline src/bin/tikv-server.rs]: 1. As far as I can see, we use `as_str` most of time. And we can pass `as_xxx` as a closure.
2. Use an `Option<T>` as default value can resolve it.

Comment (overvenus) [inline src/bin/tikv-server.rs]: PTAL @BusyJay

File: src/bin/tikv-server.rs [hunk 6]
```diff
@@ -48,13 +48,43 @@ fn print_usage(program: &str, opts: Options) {
...
+fn get_string_value<F>(short: &str,
+                       long: &str,
+                       matches: &Matches,
+                       config: &toml::Value,
+                       default: Option<String>,
+                       f: F)
+                       -> String
+    where F: Fn(&toml::Value) -> Option<String>
+{
+    matches.opt_str(short).unwrap_or_else(|| {
```
Comment (BusyJay) [inline src/bin/tikv-server.rs]: How about this?

```
matches.opt_str(short).or_else(|| {
    config.lookup(long).and_then(|v| f(&v)).or_else(|| {
        info!("malformed or missing {}, use default", long);
        default
    })
}).expect(&format!("please specify {}", long))
```

File: src/bin/tikv-server.rs [hunk 7]
```diff
@@ -48,13 +48,43 @@ fn print_usage(program: &str, opts: Options) {
...
+    })
+}
+
+fn initial_log(matches: &Matches, config: &toml::Value) {
+    let level = get_string_value("L",
+                                 "server.log",
+                                 &matches,
+                                 &config,
+                                 Some("info".to_owned()),
+                                 |v| {
```
Comment (BusyJay) [inline src/bin/tikv-server.rs]: `|v| v.map(|s| s.to_owned())` seems more compact.

**Comment by BusyJay** [comment]
rest LGTM

File: etc/config-template.toml [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+# set listening address.
+addr = "127.0.0.1:20160"
+# if not set, use addr instead. set advertise listening address for client communication.
+advertise-addr = "127.0.0.1:20161"
+# set which dsn to use, warning: default is rocksdb without persistent.
+dsn = "rocksdb"
+# set the path to rocksdb directory.
+store = "/tmp/tikv/store"
+# log level: trace, debug, info, warn, error, off.
+log = "info"
```
Comment (ngaut) [inline etc/config-template.toml]: s/log/log-level/

File: src/bin/tikv-server.rs (hunk 2 above)
Comment (ngaut) [inline src/bin/tikv-server.rs]: Use if let pattern to make code clean.
if let Some(path) = matches.opt_str("s") {
        return path;
    }

    let path = match config.lookup("server.store")

**Comment by ngaut** [comment]
LGTM

**Comment by BusyJay** [comment]
LGTM

**Comment by ngaut** [comment]
@overvenus Could you squash your commit message?

**Comment by ngaut** [comment]
I just squash it myself, thank you very much @overvenus 
Good job.

**Comment by overvenus** [comment]
@ngaut My pleasure. ;-)

**Pull request merged**